packages/contracts/contracts/SmartInvoiceFactoryV2.sol

```diff
@@ -0,0 +1,312 @@
+        require(_implementationData.length != 0, "No invoice data");
+        uint256 _implemenationVersion = getCurrentImplementationVersion(
+            _implementationType
+        );
+
+        address _implementationAddress = implementations[_implementationType][
+            _implemenationVersion
+        ];
+        require(
+            _implementationAddress != address(0),
+            "Invoice implementation does not exist"
+        );
+
+        address invoiceAddress = Clones.clone(_implementationAddress);
+
+        _init(
+            invoiceAddress,
+            _client,
+            _provider,
+            _resolverType,
+            _resolver,
+            _amounts,
+            _implementationData,
+            _implementationType,
+            _implemenationVersion,
+            _implementationAddress
+        );
+
+        return invoiceAddress;
+    }
+
+    function predictDeterministicAddress(
+        bytes32 _implementationType,
+        bytes32 _salt
+    ) external view override returns (address) {
+        uint256 _implementationVersion = getCurrentImplementationVersion(
+            _implementationType
+        );
+        return
+            Clones.predictDeterministicAddress(
+                implementations[_implementationType][_implementationVersion],
+                _salt
+            );
+    }
+
+    function createDeterministic(
+        address _client,
+        address _provider,
+        uint8 _resolverType,
+        address _resolver,
+        uint256[] calldata _amounts,
+        bytes calldata _implementationData,
+        bytes32 _implementationType,
+        bytes32 _salt
+    ) external override returns (address) {
+        // can combine this and above create into single check function maybe instead of dupes
+        require(_implementationData.length != 0, "No invoice data");
+
+        uint256 _implementationVersion = getCurrentImplementationVersion(
+            _implementationType
+        );
+        address _implementationAddress = implementations[_implementationType][
+            _implementationVersion
+        ];
+        require(
+            _implementationAddress != address(0),
+            "Invoice implementation does not exist"
+        );
+
+        address invoiceAddress = Clones.cloneDeterministic(
+            _implementationAddress,
+            _salt
+        );
+
+        _init(
+            invoiceAddress,
+            _client,
+            _provider,
+            _resolverType,
+            _resolver,
+            _amounts,
+            _implementationData,
+            _implementationType,
+            _implementationVersion,
+            _implementationAddress
+        );
+
+        return invoiceAddress;
+    }
+
+    /** @dev marks a deployed contract as a suitable implementation for additional escrow invoices formats */
+
+    function addImplementation(
+        bytes32 implementationType,
+        address implementation
+    ) external {
```

Comment:
Should include modifier for only admins to execute this function

Reply:
Yeah, this is used in 3 places. Probably makes sense to use a modifier.

Reply:
yes; had this to do on a backlog, so will do